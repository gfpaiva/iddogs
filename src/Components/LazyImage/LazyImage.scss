@import '../../Scss/root';

.lazy-image {
  @include shadow();

  &__image {
    background-color: $gray;
    background-repeat: no-repeat;
    background-size: cover;
    height: 33.333vh;
    width: 100%;
  }
}
