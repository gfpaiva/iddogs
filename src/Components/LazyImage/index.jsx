/* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import VisibilitySensor from 'react-visibility-sensor';
import { withRouter } from 'react-router-dom';
import { CSSTransition } from 'react-transition-group';

import './LazyImage.scss';
import Zoom from '../Zoom';

function LazyImage({
  image,
  id,
  history: { push },
  location: { pathname },
  match: { params },
}) {
  const [visible, setVisible] = useState(false);
  const [zoom, setZoom] = useState(false);

  useEffect(() => {
    const intId = parseInt(params.id, 10);

    if (params.id && intId === id) setZoom(true);
  }, []);

  return (
    <div className="lazy-image">
      <VisibilitySensor
        onChange={isVisible => isVisible && setVisible(true)}
        partialVisibility
        offset={{ top: 100 }}
      >
        <CSSTransition
          in={visible}
          timeout={800}
          classNames={{
            enter: 'animated fast',
            enterActive: 'fadeIn fast',
            enterDone: 'animated fadeIn fast',
          }}
        >
          <div
            className="lazy-image__image"
            onClick={() => {
              setZoom(true);
              push(`${pathname}/${id}`);
            }}
            style={{
              ...(visible) && { backgroundImage: `url(${image})` },
            }}
          />
        </CSSTransition>
      </VisibilitySensor>

      {zoom && (
        <Zoom id={id} image={image} setZoom={setZoom} />
      )}
    </div>
  );
}

LazyImage.propTypes = {
  image: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

export default withRouter(LazyImage);
