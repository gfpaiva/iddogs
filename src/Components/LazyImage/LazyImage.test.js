import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter as Router, MemoryRouter } from 'react-router-dom';''

import feed from '../../../__mocks__/getFeed';

import LazyImage from './index';

describe('<LazyImage />', () => {
  it('should mount properly', () => {
    const wrapper = mount(
    <Router>
      <LazyImage
        id={0}
        image={feed.list[0]}
      />
    </Router>
    );

    expect(wrapper.find('.lazy-image')).toMatchSnapshot();
  });

  it('should add backgroundImage when visible', () => {
    const wrapper = mount(
    <Router>
      <LazyImage
        id={0}
        image={feed.list[0]}
        forceVisible
      />
    </Router>
    );

    expect(wrapper.find('.lazy-image__image').props().style).toHaveProperty('backgroundImage');
  });

  it('should show zoom with click handler', () => {
    const wrapper = mount(
      <MemoryRouter>
        <LazyImage
          id={0}
          image={feed.list[0]}
          forceVisible
        />
      </MemoryRouter>
    );

    wrapper.find('.lazy-image__image').simulate('click');

    expect(wrapper.find('Zoom')).toHaveLength(1);
  });
});
