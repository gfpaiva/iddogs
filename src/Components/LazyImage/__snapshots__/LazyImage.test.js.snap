// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LazyImage /> should mount properly 1`] = `
<div
  className="lazy-image"
>
  <VisibilitySensor
    active={true}
    containment={null}
    delayedCall={false}
    intervalCheck={true}
    intervalDelay={100}
    minTopValue={0}
    offset={
      Object {
        "top": 100,
      }
    }
    onChange={[Function]}
    partialVisibility={true}
    resizeCheck={false}
    resizeDelay={250}
    resizeThrottle={-1}
    scrollCheck={false}
    scrollDelay={250}
    scrollThrottle={-1}
  >
    <CSSTransition
      classNames={
        Object {
          "enter": "animated fast",
          "enterActive": "fadeIn fast",
          "enterDone": "animated fadeIn fast",
        }
      }
      in={false}
      timeout={800}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={800}
        unmountOnExit={false}
      >
        <div
          className="lazy-image__image"
          onClick={[Function]}
          style={Object {}}
        />
      </Transition>
    </CSSTransition>
  </VisibilitySensor>
</div>
`;
